### ISSUE TITLE
Understanding WorkFlow for Training and Testing

### ISSUE THREAD

**Issue by lironmo** [pytorch/pytorch#6857]
hello, 
I am trying to compare all c++ framework(caffe2, tiny-dnn, etc.) with NNPACK.
the comparison that i do: Speed (exclude image loading), Memory size, and Code size.

I am new in caffe/caffe2 and worked with pytorch.

I looking a good way to learn caffe2(i already read the tutorial), the current is not sufficient.

some open issues from my side:
1. Can i stream images and labels and change the input at runtime?(do back-propagation on the mini-batch)
2. Why i need to use db input and why it's need to duplicated the data? can i avoid it?
3. How the back-propagation work? where the forward/backward implementation? when the gradients are summit? 
4. How can i insert labels for loss function on the minibatch input(i tried with tensorCPU input det as "data" and vectors of int's to "label" i get exception so it's probably not the best way)

5. Why there isn't interface like in other framework that simply run train on image:
for example in tiny- dnn i simply runed 
                        nn->train<tiny_dnn::cross_entropy>(optimizer, minibatch, labels,minibatch.size(), 1); 
where 
			std::vector<tiny_dnn::vec_t> minibatch;
			std::vector<tiny_dnn::label_t> labels;

6. I tried to switch engine to NNPACK - does it work(implement) only on arm or at intel process as well(i avoiding MKL), i get some warning that some operators are not supported - what is the fallback can i defined it?

some warning that i am not sure what im doing wrong about the input size.

I would be very happy to help!
All done in c++.

Thanks,
Liron

**Comment by jimmyoic** [comment]
1. You can load the image to your data structure such as array of float, and copy it into the input of your training net
2. like 1.
3. & 4. Back-propagation works as how you construct your network, which may consist of operators used to calculate the gradient and update the weight. Given the loss function (e.g. softmaxWithLossLayer), there are APIs to construct a training net automatically in Python. If you want to know the detail you can check data_parallel_model.py.

**Comment by lironmo** [comment]
when i tried to avoid the DB input i get wired exception(ame code with db input works):

erminate called after throwing an instance of 'caffe2::EnforceNotMet'
  what():  [enforce fail at net.cc:61] . op LabelCrossEntropy: Source for input label is unknown for net train_predict, operator input: "prob" input: "label" output: "xent" type: "LabelCrossEntropy" 
*** Aborted at 1524492241 (unix time) try "date -d @1524492241" if you are using GNU date ***
PC: @     0x7f49807ce428 gsignal
*** SIGABRT (@0x3e800006116) received by PID 24854 (TID 0x7f4968ba8b40) from PID 24854; stack trace: ***
    @     0x7f4981f15390 (unknown)
    @     0x7f49807ce428 gsignal
    @     0x7f49807d002a abort
    @     0x7f498111284d __gnu_cxx::__verbose_terminate_handler()
    @     0x7f49811106b6 (unknown)
    @     0x7f4981110701 std::terminate()
    @     0x7f4981110919 __cxa_throw
    @     0x7f498265c829 caffe2::NetBase::NetBase()
    @     0x7f498268be4a caffe2::SimpleNet::SimpleNet()
    @     0x7f498265ce36 caffe2::CreateNet()
    @     0x7f49826c30fa caffe2::Workspace::CreateNet()
    @     0x7f49826c48cf caffe2::Workspace::CreateNet()
    @           0x424ffe caffe2::Caffe2Trainer::train_as_stream()
    @           0x426a2d caffe2::Caffe2Trainer::run()
    @           0x426fa3 main
    @     0x7f49807b9830 __libc_start_main
    @           0x41cc29 _start
    @                0x0 (unknown)


it's happened before i set the "data" and labels "blob" at CreateNet

**Comment by lironmo** [comment]
i cannot use python, only cpp.

**Comment by jimmyoic** [comment]
Have you indicate the blob of label, somehow like external_input: label in predict_net?

**Comment by jimmyoic** [comment]
Would you mind to share your prototxt? (predict_net & init_net)

**Comment by lironmo** [comment]
i don't have one i can try to save it.

i am using the zoo model of resnet18

**Comment by lironmo** [comment]
layer = AddFirst("1", layer, 64, 2, train)->output(0);
      layer = AddBlock2(n, layer, 64, 64, 1, depth_2, train)->output(0);
      layer = AddBlock2(n, layer, 64, 128, 2, depth_3, train)->output(0);
      layer = AddBlock2(n, layer, 128, 256, 2, depth_4, train)->output(0);
      layer = AddBlock2(n, layer, 256, 512, 2, depth_5, train)->output(0);
      layer = AddEnd(tos(n++), layer, 512, out_size, 1)->output(0);

**Comment by jimmyoic** [comment]
To generate the training prototxt correctly, I'll suggest you use model helper(python) to build it. You can still run it on C++

**Comment by lironmo** [comment]
In general i prefer to use cpp code only because i will need to create the net at runtime at cpp.

But i will happy to hear how to create the net correctly and find out what is the problem with my implementation.

I used caffe2 cpp tutorial, i also take a look on your demo code to see if i can use it in similar  way

**Comment by lironmo** [comment]
Hi i changed the order of the creation and mutation of the "label" "data" blobs and now i get the following errors:
terminate called after throwing an instance of 'caffe2::EnforceNotMet'
  what():  [enforce fail at net.cc:61] . op Adam: Source for input label_grad is unknown for net train_predict, operator input: "label" input: "label_moment_1" input: "label_moment_2" input: "label_grad" input: "lr" input: "iter" output: "label" output: "label_moment_1" output: "label_moment_2" type: "Adam" 
*** Aborted at 1524558114 (unix time) try "date -d @1524558114" if you are using GNU date ***

i added the following:
      train_model.predict.net.add_external_input("label_grad");

and get the following error:
terminate called after throwing an instance of 'caffe2::EnforceNotMet'
  what():  [enforce fail at operator.cc:42] blob != nullptr. op Adam: Encountered a non-existing input blob: label_grad 

why i need to feed this blob as well?

same code with minor change (using DB input ) work fine:
      train_model.AddDatabaseOps("train", model_in, db_paths[0],
                                 db_type, batch_size);

so i am not aware to all the blobs that need to be inserted , my code based on /leonardvandriel/caffe2_cpp_tutorial , and i simply tried to remove the DB input and set the labels and data for each interrelation in order to stream the imaged like in https://caffe2.ai/docs/intro-tutorial.html at cpp.

in regards to the NNPACK (6) , my workaround is to set it only to the convolution op:
  void NetUtil::SetConvEngine(const std::string engine) {
    for (auto &op : *net.mutable_op()) {
      if ("Conv" == op.type()) {
        op.set_engine(engine);
      }
    }
  }

1 +2 - is still open and i will happy for help(simply remove the DB input to stream the images one by one or up to the mini batch - and do backward pass after finish with all the mini batch)
